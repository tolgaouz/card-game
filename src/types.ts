import { TransitionStatus } from 'react-transition-group';

export enum CardKind {
  CLUBS = 'Clubs',
  DIAMONDS = 'Diamonds',
  HEARTS = 'Hearts',
  SPADES = 'Spades',
}
/*
Denote non-number types with numbers as well,
1 --> Ace
11 --> Jack
12 --> Queen
13 --> King
*/
export type CardNumber = 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12 | 13;
export enum CardKindColorCSS {
  RED = 'text-redCard',
  BLACK = 'text-black',
}
export type Card = { kind: CardKind; number: CardNumber };

export type Ace = { number: 1; kind: CardKind };
export type King = { number: 13; kind: CardKind };
export type Queen = { number: 12; kind: CardKind };
export type Jack = { number: 11; kind: CardKind };

export interface CardGameState {
  deck: Card[];
  hand: Card[];
  acesInDeck: Ace[];
  finished?: boolean;
  userWon?: boolean;
}

export type TransitionStyle = Exclude<TransitionStatus, 'unmounted'>;
