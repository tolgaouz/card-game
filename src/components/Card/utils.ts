import ClubsSVG from '../../assets/Clubs.svg';
import DiamondSVG from '../../assets/Diamonds.svg';
import HeartSVG from '../../assets/Hearts.svg';
import SpadeSVG from '../../assets/Spades.svg';
import { CardNumber, CardKind, CardKindColorCSS } from '../../types';

export const errorMessages = {
  cardKind: 'Type parameter needs to be one of CardType enums.',
  cardNumber: 'Card number can not be more than 13',
};

export const getCardColor = (type: CardKind): CardKindColorCSS => {
  if ([CardKind.CLUBS, CardKind.SPADES].includes(type)) return CardKindColorCSS.BLACK;
  if ([CardKind.DIAMONDS, CardKind.HEARTS].includes(type)) return CardKindColorCSS.RED;
  throw new Error(errorMessages.cardKind);
};

export const getCardText = (number: CardNumber): string => {
  switch (number) {
    case 1:
      return 'Ace';
    case 11:
      return 'Jack';
    case 12:
      return 'Queen';
    case 13:
      return 'King';
    default:
      if (number > 13) throw new Error(errorMessages.cardNumber);
      return number.toString();
  }
};

export const getCardSymbol = (type: CardKind): string => {
  switch (type) {
    case CardKind.CLUBS:
      return ClubsSVG;
    case CardKind.DIAMONDS:
      return DiamondSVG;
    case CardKind.HEARTS:
      return HeartSVG;
    case CardKind.SPADES:
      return SpadeSVG;
    default:
      throw new Error(errorMessages.cardKind);
  }
};
