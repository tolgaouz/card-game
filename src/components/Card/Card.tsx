import clsx from 'clsx';
import React from 'react';
import { CardKind, CardNumber } from '../../types';
import { getCardText, getCardColor, getCardSymbol } from './utils';

interface CardProps extends React.HTMLProps<HTMLDivElement> {
  kind: CardKind;
  number: CardNumber;
}

const Card: React.FC<CardProps> = ({ kind, number, className, ...restProps }) => {
  const firstLetter = getCardText(number).slice(0)[0];
  const symbol = getCardSymbol(kind);
  const text = `${getCardText(number)} of ${kind}`;
  const color = getCardColor(kind);
  return (
    <div
      className={clsx(
        'bg-white rounded-lg md:rounded-2xl lg:rounded-3xl relative font-courier w-1/6 min-w-[103px] max-w-[192px] aspect-[2.2/3]',
        className
      )}
      {...restProps}
    >
      <div
        className={clsx(
          'font-bold leading-none relative w-fit text-4xl md:text-6xl lg:text-8xl pt-5 pl-5',
          color
        )}
        data-testid="card-text"
      >
        {firstLetter.toUpperCase()}
        <img
          className="absolute min-w-[20px] aspect-[1/1] max-w-[36px] w-1/2 right-0 lg:right-1 lg:-bottom-6"
          src={symbol}
          alt={text}
        />
      </div>
      <img
        data-testid="card-symbol"
        className="min-w-[48px] max-w-[90px] w-1/2 aspect-[1/1] absolute right-3.5 bottom-3.5 lg:right-6 lg:bottom-6"
        src={symbol}
        alt={text}
      />
    </div>
  );
};

export default Card;
