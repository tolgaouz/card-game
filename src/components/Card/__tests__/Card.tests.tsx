import { render } from '@testing-library/react';
import { Card as CardT, CardKind, Ace, King, Jack, Queen } from '../../../types';
import Card from '../Card';

const Aces: Ace[] = Object.values(CardKind).map((kind) => ({ kind, number: 1 }));
const Kings: King[] = Object.values(CardKind).map((kind) => ({ kind, number: 13 }));
const Queens: Queen[] = Object.values(CardKind).map((kind) => ({ kind, number: 12 }));
const Jacks: Jack[] = Object.values(CardKind).map((kind) => ({ kind, number: 11 }));
const Numbers: CardT[] = Object.values(CardKind).map((kind) => ({ kind, number: 3 }));

const Cards = [Aces, Kings, Queens, Jacks, Numbers];

describe('Card', () => {
  it('renders correctly', () => {
    const { container } = render(
      <div>
        {Cards.map(([card]) => (
          <Card key={`${card.kind}_${card.number}`} {...card} />
        ))}
      </div>
    );
    expect(container).toMatchSnapshot();
  });

  it('renders all kinds of standard cards properly', () => {
    const { queryAllByTestId, queryAllByAltText } = render(
      <div>
        {Numbers.map((card) => (
          <Card key={`${card.kind}_${card.number}`} {...card} />
        ))}
      </div>
    );
    Object.values(CardKind).forEach((kind) => {
      const symbolImages = queryAllByAltText(`3 of ${kind}`);
      expect(symbolImages[1].getAttribute('src')).toBe(`${kind}.svg`);
      queryAllByTestId('card-text').forEach((node) => {
        expect(node.textContent).toBe('3');
      });
      expect(queryAllByAltText(`3 of ${kind}`).length).toBe(2);
    });
  });

  it('renders all kinds of ace cards properly', () => {
    const { queryAllByTestId, queryAllByAltText } = render(
      <div>
        {Aces.map((card) => (
          <Card key={`${card.kind}_${card.number}`} {...card} />
        ))}
      </div>
    );
    Object.values(CardKind).forEach((kind) => {
      const symbolImages = queryAllByAltText(`Ace of ${kind}`);
      expect(symbolImages[1].getAttribute('src')).toBe(`${kind}.svg`);
      queryAllByTestId('card-text').forEach((node) => {
        expect(node.textContent).toBe('A');
      });
      expect(queryAllByAltText(`Ace of ${kind}`).length).toBe(2);
    });
  });

  it('renders all kinds of jack cards properly', () => {
    const { queryAllByTestId, queryAllByAltText } = render(
      <div>
        {Jacks.map((card) => (
          <Card key={`${card.kind}_${card.number}`} {...card} />
        ))}
      </div>
    );
    Object.values(CardKind).forEach((kind) => {
      const symbolImages = queryAllByAltText(`Jack of ${kind}`);
      expect(symbolImages[1].getAttribute('src')).toBe(`${kind}.svg`);
      queryAllByTestId('card-text').forEach((node) => {
        expect(node.textContent).toBe('J');
      });
      expect(queryAllByAltText(`Jack of ${kind}`).length).toBe(2);
    });
  });

  it('renders all kinds of queen cards properly', () => {
    const { queryAllByTestId, queryAllByAltText } = render(
      <div>
        {Queens.map((card) => (
          <Card key={`${card.kind}_${card.number}`} {...card} />
        ))}
      </div>
    );
    Object.values(CardKind).forEach((kind) => {
      const symbolImages = queryAllByAltText(`Queen of ${kind}`);
      expect(symbolImages[1].getAttribute('src')).toBe(`${kind}.svg`);
      queryAllByTestId('card-text').forEach((node) => {
        expect(node.textContent).toBe('Q');
      });
      expect(queryAllByAltText(`Queen of ${kind}`).length).toBe(2);
    });
  });

  it('renders all kinds of king cards properly', () => {
    const { queryAllByTestId, queryAllByAltText } = render(
      <div>
        {Kings.map((card) => (
          <Card key={`${card.kind}_${card.number}`} {...card} />
        ))}
      </div>
    );
    Object.values(CardKind).forEach((kind) => {
      const symbolImages = queryAllByAltText(`King of ${kind}`);
      expect(symbolImages[1].getAttribute('src')).toBe(`${kind}.svg`);
      queryAllByTestId('card-text').forEach((node) => {
        expect(node.textContent).toBe('K');
      });
      expect(symbolImages.length).toBe(2);
    });
  });
});
