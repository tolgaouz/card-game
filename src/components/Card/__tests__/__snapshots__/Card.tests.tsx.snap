// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card renders correctly 1`] = `
<div>
  <div>
    <div
      class="bg-white rounded-lg md:rounded-2xl lg:rounded-3xl relative font-courier w-1/6 min-w-[103px] max-w-[192px] aspect-[2.2/3]"
    >
      <div
        class="font-bold leading-none relative w-fit text-4xl md:text-6xl lg:text-8xl pt-5 pl-5 text-black"
        data-testid="card-text"
      >
        A
        <img
          alt="Ace of Clubs"
          class="absolute min-w-[20px] aspect-[1/1] max-w-[36px] w-1/2 right-0 lg:right-1 lg:-bottom-6"
          src="Clubs.svg"
        />
      </div>
      <img
        alt="Ace of Clubs"
        class="min-w-[48px] max-w-[90px] w-1/2 aspect-[1/1] absolute right-3.5 bottom-3.5 lg:right-6 lg:bottom-6"
        data-testid="card-symbol"
        src="Clubs.svg"
      />
    </div>
    <div
      class="bg-white rounded-lg md:rounded-2xl lg:rounded-3xl relative font-courier w-1/6 min-w-[103px] max-w-[192px] aspect-[2.2/3]"
    >
      <div
        class="font-bold leading-none relative w-fit text-4xl md:text-6xl lg:text-8xl pt-5 pl-5 text-black"
        data-testid="card-text"
      >
        K
        <img
          alt="King of Clubs"
          class="absolute min-w-[20px] aspect-[1/1] max-w-[36px] w-1/2 right-0 lg:right-1 lg:-bottom-6"
          src="Clubs.svg"
        />
      </div>
      <img
        alt="King of Clubs"
        class="min-w-[48px] max-w-[90px] w-1/2 aspect-[1/1] absolute right-3.5 bottom-3.5 lg:right-6 lg:bottom-6"
        data-testid="card-symbol"
        src="Clubs.svg"
      />
    </div>
    <div
      class="bg-white rounded-lg md:rounded-2xl lg:rounded-3xl relative font-courier w-1/6 min-w-[103px] max-w-[192px] aspect-[2.2/3]"
    >
      <div
        class="font-bold leading-none relative w-fit text-4xl md:text-6xl lg:text-8xl pt-5 pl-5 text-black"
        data-testid="card-text"
      >
        Q
        <img
          alt="Queen of Clubs"
          class="absolute min-w-[20px] aspect-[1/1] max-w-[36px] w-1/2 right-0 lg:right-1 lg:-bottom-6"
          src="Clubs.svg"
        />
      </div>
      <img
        alt="Queen of Clubs"
        class="min-w-[48px] max-w-[90px] w-1/2 aspect-[1/1] absolute right-3.5 bottom-3.5 lg:right-6 lg:bottom-6"
        data-testid="card-symbol"
        src="Clubs.svg"
      />
    </div>
    <div
      class="bg-white rounded-lg md:rounded-2xl lg:rounded-3xl relative font-courier w-1/6 min-w-[103px] max-w-[192px] aspect-[2.2/3]"
    >
      <div
        class="font-bold leading-none relative w-fit text-4xl md:text-6xl lg:text-8xl pt-5 pl-5 text-black"
        data-testid="card-text"
      >
        J
        <img
          alt="Jack of Clubs"
          class="absolute min-w-[20px] aspect-[1/1] max-w-[36px] w-1/2 right-0 lg:right-1 lg:-bottom-6"
          src="Clubs.svg"
        />
      </div>
      <img
        alt="Jack of Clubs"
        class="min-w-[48px] max-w-[90px] w-1/2 aspect-[1/1] absolute right-3.5 bottom-3.5 lg:right-6 lg:bottom-6"
        data-testid="card-symbol"
        src="Clubs.svg"
      />
    </div>
    <div
      class="bg-white rounded-lg md:rounded-2xl lg:rounded-3xl relative font-courier w-1/6 min-w-[103px] max-w-[192px] aspect-[2.2/3]"
    >
      <div
        class="font-bold leading-none relative w-fit text-4xl md:text-6xl lg:text-8xl pt-5 pl-5 text-black"
        data-testid="card-text"
      >
        3
        <img
          alt="3 of Clubs"
          class="absolute min-w-[20px] aspect-[1/1] max-w-[36px] w-1/2 right-0 lg:right-1 lg:-bottom-6"
          src="Clubs.svg"
        />
      </div>
      <img
        alt="3 of Clubs"
        class="min-w-[48px] max-w-[90px] w-1/2 aspect-[1/1] absolute right-3.5 bottom-3.5 lg:right-6 lg:bottom-6"
        data-testid="card-symbol"
        src="Clubs.svg"
      />
    </div>
  </div>
</div>
`;
